typo: fix missing backtick (#20202)
cloudflare/cloudflare-docs: src/content/docs/agents/examples/manage-and-sync-state.mdx

---
title: Manage and sync state
pcx_content_type: concept
sidebar:
  order: 6
---

import { MetaInfo, Render, Type, TypeScriptExample, WranglerConfig } from "~/components";

Every Agent has built-in state management capabilities, including built-in storage and synchronization between the Agent and frontend applications. State within an Agent is:

* Persisted across Agent restarts: data is permanently persisted within the Agent.
* Automatically serialized/deserialized: you can store any JSON-serializable data.
* Immediately consistent within the Agent: read your own writes.
* Thread-safe for concurrent updates

Agent state is stored in SQL database that associate with each indidivual Agent instance: you can interact with it using the higher-level `this.setState` API (recommended) or by directly querying the database with `this.sql`.

### State API

> Every agent has built-in state management capabilities. You can set and update the agent's state directly using `this.setState`:

Reviewer: ```suggestion
Every Agent has built-in state management capabilities. You can set and update the Agent's state directly using `this.setState`:
```
I believe we're talking about the CF product "Agent" in this chapter. Capitalising Agents throughout.

<TypeScriptExample>

```ts
import { Agent } from "@cloudflare/agents";

export class MyAgent extends Agent {
  // Update state in response to events
  async incrementCounter() {
    this.setState({
      ...this.state,
      counter: this.state.counter + 1,
    });
  }

  // Handle incoming messages
  async onMessage(message) {
    if (message.type === "update") {
      this.setState({
        ...this.state,
        ...message.data,
      });
    }
  }

  // Handle state updates
  onStateUpdate(state, source: "server" | Connection) {
    console.log("state updated", state);
  }
}
```

</TypeScriptExample>

> If you're using TypeScript, you can also provide a type for your Agent's state by passing in a type as a [type parameter](https://www.typescriptlang.org/docs/handbook/2/generics.html#using-type-parameters-in-generic-constraints) as the _second_ type parameter to the `Agent` class definition.

Reviewer: ```suggestion
If you are using TypeScript, you can also provide a type for your Agent's state by passing in a type as a [type parameter](https://www.typescriptlang.org/docs/handbook/2/generics.html#using-type-parameters-in-generic-constraints) as the _second_ type parameter to the `Agent` class definition.
```

<TypeScriptExample>

```ts
import { Agent } from "@cloudflare/agents";

interface Env {}

//
interface FlightRecord {
	id: string;
	departureIata: string;
	arrival: Date;;
	arrivalIata: string;
	price: number;
}

// Pass in the type of your Agent's state
export class MyAgent extends Agent<Env, FlightRecord> {
  // This allows this.setState and
}
```

</TypeScriptExample>

### Synchronizing state

Clients can connect to an Agent and stay synchronized with its state using the React hooks provided as part of `@cloudflare/agents/react`:

<TypeScriptExample>

```ts

> 

Reviewer: ```suggestion
TODO
```
Flagging empty TS codeblock in case it flies under the radar.

```

</TypeScriptExample>

The state synchronization system:

> * Automatically syncs the agent's state to all connected clients

Reviewer: ```suggestion
* Automatically syncs the Agent's state to all connected clients
```

> * Handles client disconnections and reconnections gracefully

Reviewer: ```suggestion
* Handles client disconnections and reconnections
```

* Provides immediate local updates
* Supports multiple simultaneous client connections

Common use cases:

* Real-time collaborative features
* Multi-window/tab synchronization
* Live updates across multiple devices
* Maintaining consistent UI state across clients

> * When new clients connect, they automatically receive the current state from the agent, ensuring all clients start with the latest data.

Reviewer: ```suggestion
* When a new client connects, they automatically receive the current state from the Agent, ensuring all clients start with the latest data.
```
